### ISSUE TITLE
Render from opengl FBO

### ISSUE THREAD

**Issue by vgarleanu** [slint-ui/slint#423]
I would like to add support for a video widget using [`libmpv`.](https://github.com/mpv-player/mpv/tree/master/libmpv) [`libmpv`](https://github.com/mpv-player/mpv/tree/master/libmpv) can display the video overlayed on top of the parent window, but it is preferred to have [`libmpv`](https://github.com/mpv-player/mpv/tree/master/libmpv) draw into a FBO which you then render to the screen.

This library doesnt seem to support rendering from gl FBOs. How difficult would it be to add such a feature in?

**Comment by tronical** [comment]
Welcome back from Reddit ;-). Yes, I think supporting this makes sense.

There are two considerations that I think apply:

1. We support different rendering backends, and only one of them is OpenGL based. It's fine to offer functionality that only works with one rendering backend, but we need to make it easy for applications to find out if a certain feature is supported and be able to fall back if necessary. Similarly, we may need a way for applications to explicitly choose a backend - in your case that might make sense. This could be a capability based query API, for example.
2. I think the API to use here is `sixtyfps::Image`. Currently that can be constructed from a path or a pixel buffer, but for your use-case I think we would want the ability to create an image from a texture id - like a `u32`. That in turn requires API to allow for you to create an OpenGL context that's sharing resources with the OpenGL context that the OpenGL based rendering backend is using.

**Comment by vgarleanu** [comment]
I like the idea of being able to construct an Image from a texture id. That should be sufficient for my use case. But perhaps it would make more sense to offer some sort of canvas API.

Also, it seems that `femtovg` has support for rendering into a framebuffer object, but I can't seem to find any APIs to actually render the framebuffer objects, is that something that sixtyfps is supposed to handle separately?

**Comment by tronical** [comment]
You're right, using femtovg directly with an FBO might just be fine. You'd basically create a texture yourself and set that as a screen target on femtovg's OpenGL renderer. In order for us to implement support for rendering that texture (`sixtysp::Image::from_opengl_texture`) we need to extend femtovg with the ability to kind of "adopt" a texture as a `femtovg::Image`.

**Comment by vgarleanu** [comment]
I see, I will open a issue regarding creating a image from a texture on the `femtovg` repository. 

Regarding sharing the opengl/glow context, what do you think the best approach is, api-wise? Since the context is an `Rc` we could just have a method that clones it.

**Comment by tronical** [comment]
> I see, I will open a issue about creating a image from a texture on the `femtovg` repository.

👍 

> Regarding sharing the opengl/glow context, what do you think the best approach is, api-wise? Since the context is an `Rc` we could just have a method that clones it.

That's a good question. With sharing I meant sharing resources in the OpenGL sense. So API wise this could be as minimal as the GL backend offering an API to create it and a way to make the context current. Then in C-land I'd ask the GL library for the current context of the thread and use that to create a new sharing context. I need to see how that could be done with glutin :)

**Comment by tronical** [comment]
Related issue in glutin: rust-windowing/glutin#1337

**Comment by tronical** [comment]
Another use-case for this is embedding of opengl mapbox.

**Comment by sztomi** [comment]
The core performance feature of mpv is that it doesn't do an unnecessary round-trip between the CPU and the GPU (all of the rendering is done on the GPU). Rasterizing the opengl texture into an in-memory `Image` and then returning that to the GPU is very slow. This is why libmpv recommends using either its opengl or the native window embedding API. It does have a bitmap output so in order to use that with sixtyfps there is no extra work needed in this library.

**Comment by tronical** [comment]
Ah. This could be a misunderstanding. The idea with this ticket is *not* to download the GPU rasterized image but allow the `Image` to retain a handle to the texture.

**Comment by sztomi** [comment]
Ah I see, I indeed misunderstood that. In that case my comment is moot. Thanks for pointing that out.

FWIW I am the author of the linked issue in glutin. Later on I ended up working out the opengl integration in iced and upstreamed my solution in a PR: https://github.com/iced-rs/iced/blob/master/examples/integration_opengl/src/main.rs 

tl;dr is basically creating the context outside of both the GUI library and the scene rendered, and then providing each with the handle to render and window dimensions to render one after the other. I'm not sure if the currenty sixtyfps API has extension points for this, but I think this would also be a reasonable approach for allowing low-level integration (i.e. not just for content that is rendered to an opengl texture but anything at all). The `Image` approach also sounds great for libmpv integration particularly if it's possible to implement.

**Comment by vgarleanu** [comment]
Given that support for OpenGL underlays has been merged in, I believe this can now be achieved with underlays.

**Comment by sztomi** [comment]
@vgarleanu can you point me towards documentation or an example of that?

**Comment by vgarleanu** [comment]
> @vgarleanu can you point me towards documentation or an example of that?

Hey, I've just written a pretty rough demo of getting this with Flint, you can access it here: https://github.com/vgarleanu/lumiere/blob/master/src/main.rs

The code is imo pretty bad but it kinda works. At the moment the main issue is that I have no clue how to set the FBO to take the full height and width of the window, and for MPV to render at that height and width. So at the moment the buffer size are set to 1920x1080.

**Comment by tronical** [comment]
https://github.com/slint-ui/slint/pull/2767 adds API in Rust and C++ to borrow OpenGL textures, and it comes with an example that renders a cube into an FBO and then blends it with Slint controls.
